packages/platforms/mellanox/x86-64/x86-64-mlnx-msn2700/onlp/builds/src/module/src/sfpi.c

```diff
@@ -0,0 +1,198 @@
+
+    if (ret == 0) {
+        if (!strncmp(buf, SFP_PRESENT_STATUS, sizeof(buf))) {
+            *value = 1;
+        } else if (!strncmp(buf, SFP_NOT_PRESENT_STATUS, sizeof(buf))) {
+            *value = 0;
+        }
+    }
+
+    return ret;
+}
+
+static char*
+sn2700_sfp_get_port_path(int port, char *node_name)
+{
+    sprintf(sfp_node_path, "/bsp/qsfp/qsfp%d%s", port, node_name);
+    return sfp_node_path;
+}
+
+/************************************************************
+ *
+ * SFPI Entry Points
+ *
+ ***********************************************************/
+int
+onlp_sfpi_init(void)
+{
+    /* Called at initialization time */
+    return ONLP_STATUS_OK;
+}
+
+int
+onlp_sfpi_bitmap_get(onlp_sfp_bitmap_t* bmap)
+{
+    /*
+     * Ports {1, 32}
+     */
+    int p = 1;
+    AIM_BITMAP_CLR_ALL(bmap);
+
+    for (; p <= NUM_OF_SFP_PORT; p++) {
+        AIM_BITMAP_SET(bmap, p);
+    }
+
+    return ONLP_STATUS_OK;
+}
+
+int
+onlp_sfpi_is_present(int port)
+{
+    /*
+     * Return 1 if present.
+     * Return 0 if not present.
+     * Return < 0 if error.
+     */
+    int present = -1;
+    char* path = sn2700_sfp_get_port_path(port, "_status");
+
+    if (sn2700_sfp_node_read_int(path, &present, 0) != 0) {
+        AIM_LOG_ERROR("Unable to read present status from port(%d)\r\n", port);
+        return ONLP_STATUS_E_INTERNAL;
+    }
+
+    return present;
+}
+
+int
+onlp_sfpi_presence_bitmap_get(onlp_sfp_bitmap_t* dst)
+{
+    int ii = 1;
+    int rc = 0;
+
+    for (;ii <= NUM_OF_SFP_PORT; ii++) {
+        rc = onlp_sfpi_is_present(ii);
+        AIM_BITMAP_MOD(dst, ii, (1 == rc) ? 1 : 0);
+    }
+
+    return ONLP_STATUS_OK;
+}
+
+int
+onlp_sfpi_eeprom_read(int port, uint8_t data[256])
+{
+    char* path = sn2700_sfp_get_port_path(port, "");
+
+    /*
+     * Read the SFP eeprom into data[]
+     *
+     * Return MISSING if SFP is missing.
+     * Return OK if eeprom is read
+     */
+    memset(data, 0, 256);
+
+    if (deviceNodeReadBinary(path, (char*)data, 256, 256) != 0) {
+        AIM_LOG_ERROR("Unable to read eeprom from port(%d)\r\n", port);
+        return ONLP_STATUS_E_INTERNAL;
+    }
+
+    return ONLP_STATUS_OK;
+}
+
+int
+onlp_sfpi_dom_read(int port, uint8_t data[256])
+{
+    return onlp_sfpi_eeprom_read(port, data);
+}
+
+int
+onlp_sfpi_dev_readb(int port, uint8_t devaddr, uint8_t addr)
+{
+    return ONLP_STATUS_E_UNSUPPORTED;
+}
+
+int
+onlp_sfpi_dev_writeb(int port, uint8_t devaddr, uint8_t addr, uint8_t value)
+{
+    return ONLP_STATUS_E_UNSUPPORTED;
+}
+
+int
+onlp_sfpi_dev_readw(int port, uint8_t devaddr, uint8_t addr)
+{
+    return ONLP_STATUS_E_UNSUPPORTED;
+}
+
+int
+onlp_sfpi_dev_writew(int port, uint8_t devaddr, uint8_t addr, uint16_t value)
+{
+    return ONLP_STATUS_E_UNSUPPORTED;
+}
+
+int
+onlp_sfpi_control_supported(int port, onlp_sfp_control_t control, int* rv)
+{
+    return ONLP_STATUS_E_UNSUPPORTED;
+}
+
+int
+onlp_sfpi_control_get(int port, onlp_sfp_control_t control, int* value)
```

Comment:
This needs to be implemented for proper operation. For QSFP{28} the only relevant controls will be RESET and LP_MODE. 

You also need onlp_sfpi_control_set(), which is missing.